### DISCUSSION TITLE
Multiple application instances

### DISCUSSION THREAD

**Discussion by calebboyd** [sequelize/umzug#59] [Ideas]
We're getting ready to deploy a schema using umzug for migrations. I was wondering (couldn't find in the docs) if the logs table acted as a mutex in the event the applications were scaled horizontally in front of a single database. eg. ensuring that `my-migration.js` was only started and completed once even if multiple servers run it against the same database.

**Comment by sdepold** [comment]
Hi @calebboyd,

if you are running the migrations on two different machines in parallel (or even on one machine twice), you will probably have awkward times, since the migration is only marked as executed, when it finished. So you should make sure to only run the migration once

**Comment by jukkah** [comment]
@sdepold I'm wondering if migrations should be executed and (un)logged in a single transaction unit or two if storage != DB?

**Comment by sdepold** [comment]
I think I tried that at some point and noticed that for example structural changes cannot be wrapped (at least not for every db) in a transaction. Can you elaborate on that idea @jukkah ?

**Comment by jukkah** [comment]
@sdepold You are right, it will not work. At least, in MySQL, DDL statements cannot be rolled back. @calebboyd's problem can be still resolved by locking the storage (file or table) as he originally suggested.

**Reply by kasir-barati** [reply]
Do you have a repo where you're locking the DB before executing the migration?

**Comment by gvko** [comment]
@calebboyd we were facing the same issue and a workaround we found was:
1. have some flag that the application is/is not a MIGRATION_HOST
2. upon deploy run in the following steps:
   * upgrade the app as a `MIGRATION_HOST=true` (which runs the migrations and updates the DB)
   * kill the app after the migrations executions have finished
   * upgrade the rest of the instances of the as `MIGRATION_HOST=false`

The whole process is automated. :)

**Comment by soulchild** [comment]
Maybe we can make this work by locking the `SequelizeMeta` table `FOR UPDATE`? 

The first process running the migrations will issue a `SELECT * FROM "SequelizeMeta" FOR UPDATE` query which basically makes all other processes wait until the table has been updated by the first process, seeing the **updated** data afterwards. 

I don't know about other databases, but I think there are equivalent solutions for most of them. At least, I believe that's how Flyway tackles this issue. There are [several implementations for all supported databases](https://github.com/flyway/flyway/search?utf8=%E2%9C%93&q=dolock) of the `doLock()` method which seems to be responsible for locking Flyway's meta table.

**Reply by gvko** [reply]
check my comment: https://github.com/sequelize/umzug/discussions/59#discussioncomment-1562923

**Comment by ababich** [comment]
> Maybe we can make this work by locking the `SequelizeMeta` table `FOR UPDATE`?
> 
> The first process running the migrations will issue a `SELECT * FROM "SequelizeMeta" FOR UPDATE` query which basically makes all other processes wait until the table has been updated by the first process, seeing the **updated** data afterwards.
> 
> I don't know about other databases, but I think there are equivalent solutions for most of them. At least, I believe that's how Flyway tackles this issue. There are [several implementations for all supported databases](https://github.com/flyway/flyway/search?utf8=%E2%9C%93&q=dolock) of the `doLock()` method which seems to be responsible for locking Flyway's meta table.

migrations table lock is used successfully in rails migrations for years and I'm truly shocked it is not even an option in Sequelize

**Comment by SamSpiri** [comment]
I just got this problem as a devops. it's really the wtf moment here. where are the locks? why nodejs comunity is like that most of the time... Gone back into my cave to build another shit tech workaround with some sticks and sticky tape, as I don't know how to code that dumb nodejs...

**Reply by gvko** [reply]
check my comment: https://github.com/sequelize/umzug/discussions/59#discussioncomment-1562923

**Comment by gvko** [comment]
Alright, guys, finally it's here: [db-migrator-distributed](https://www.npmjs.com/package/db-migrator-distributed). For the lack of a better name, that's what I chose... :D 

Enjoy, give feedback and contribute! 🙌

**Reply by kasir-barati** [reply]
Do you know of any lib for MongoDB?

**Reply by gvko** [reply]
you mean a lib that handles distributed migrations with Mongo? Not really.
But this shouldn't be to hard to add to my lib. 
What are you using for connecting to Mongo? `mongoose` or smth else?

**Reply by kasir-barati** [reply]
Mongoose.

**Reply by gvko** [reply]
Alright, I'll try to find some free time after work in the following days to add MongoDB. I haven't worked much with Mongo, so there's that as well... 
So, I hope I'll manage to add it in the next 2 weeks. :)

**Reply by kasir-barati** [reply]
Fantastic, I'll be glad to give you a hand in case you needed it :smile:.
